h2/src/main/org/h2/expression/ExpressionColumn.java

```diff
@@ -102,19 +102,14 @@ public void mapColumns(ColumnResolver resolver, int level, int state) {
             }
             return;
         }
-        for (Column col : resolver.getColumns()) {
-            String n = resolver.getColumnName(col);
-            if (database.equalsIdentifiers(columnName, n)) {
-                mapColumn(resolver, col, level);
-                if (resolver.hasDerivedColumnList()) {
-                    columnName = n;
-                }
```

Comment:
if above chunk of code seems to have gone missing?

Reply:
Oh, I missed a very specific case: if identifiers aren't case-sensitive (non-default configuration), we need to set `columnName` to original name. With default settings they are always equal to each other.

Reply:
On the other hand, without a derived column list the name isn't updated too.

Reply:
It means that H2 have a minor problem with `ResultSetMetaData.getColumnName()` with some settings.

```Java
try (Connection c = DriverManager.getConnection(
       "jdbc:h2:mem:1;DATABASE_TO_UPPER=FALSE;CASE_INSENSITIVE_IDENTIFIERS=TRUE")) {
    Statement s = c.createStatement();
    s.execute("CREATE TABLE TEST(A INT)");
    ResultSet rs = s.executeQuery("SELECT a from test");
    ResultSetMetaData md = rs.getMetaData();
    System.out.println(md.getTableName(1) + '.' + md.getColumnName(1)
            + " AS " + md.getColumnLabel(1));
}
```
```
TEST.a AS A
```

Reply:
Their real names should be returned now.